### ISSUE TITLE
Unrecoverable sync issue where MobileServicePushFailedException has zero errors

### ISSUE THREAD

**Issue by ronnyhash** [Azure/azure-mobile-apps-net-client#442]
I have a number of users getting into a state where PushAsync generates a MobileServicePushFailedException, but there are no PushResult.Errors. Since there are no errors to loop through, it is impossible to cancel and discard the update. Therefore, every attempt to push the changes fails.

This is because the call to RemoveSystemPropertiesKeepVersion() on line 174 in PushAction.cs is not in the try/catch block that begins on line 178:
operation.Item = MobileServiceSyncTable.RemoveSystemPropertiesKeepVersion(operation.Item);
https://github.com/Azure/azure-mobile-apps-net-client/blob/master/src/Microsoft.Azure.Mobile.Client/Table/Sync/Queue/Actions/PushAction.cs

A null reference exception is being generated in MobileServiceSerializer.RemoveSystemProperties which is ultimately called from the line mentioned above. You can see this in the stack trace included below.

This results in the following unrecoverable situation
1 - When the exception occurs, the ExecuteOperationAsync() method in PushAction no longer continues execution.
2 - Because there is also no try/catch in ExecuteAllOperationsAsync() which calls ExecuteOperationAsync(), that method no longer continues to execute. This is the major problem because no other pending operations will get processed.
3 - Finally, ExecuteBatchAsync() catches the exception. However, it does not add it to syncErrors.

So, essentially, the one operation that has a problem blocks any other pending operations from getting pushed. This is catastrophic! There is no way to recover.

This is a critical and major bug. Please fix this fast! I have production users ready to abandon my app.

The error is:
Push operation has failed. See the PushResult for details.
Microsoft.WindowsAzure.MobileServices.Sync.MobileServicePushFailedException

The inner exception is:
Object reference not set to an instance of an object

The stack trace from Azure App Center:
Microsoft.WindowsAzure.MobileServices
MobileServiceSerializer.RemoveSystemProperties (Newtonsoft.Json.Linq.JObject instance, System.String& version, Microsoft.WindowsAzure.MobileServices.MobileServiceSystemProperties propertiesToKeep)
Microsoft.WindowsAzure.MobileServices.Sync
MobileServiceSyncTable.RemoveSystemPropertiesKeepVersion (Newtonsoft.Json.Linq.JObject instance)
Microsoft.WindowsAzure.MobileServices.Sync
PushAction+<ExecuteOperationAsync>d__10.MoveNext ()
System.Runtime.ExceptionServices
ExceptionDispatchInfo.Throw ()
System.Runtime.CompilerServices
TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task)
System.Runtime.CompilerServices
TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task)
System.Runtime.CompilerServices
TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task)
System.Runtime.CompilerServices
TaskAwaiter`1[TResult].GetResult ()
Microsoft.WindowsAzure.MobileServices.Sync
PushAction+<ExecuteAllOperationsAsync>d__9.MoveNext ()
System.Runtime.ExceptionServices
ExceptionDispatchInfo.Throw ()
System.Runtime.CompilerServices
TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task)
System.Runtime.CompilerServices
TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task)
System.Runtime.CompilerServices
TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task)
System.Runtime.CompilerServices
TaskAwaiter.GetResult ()
Microsoft.WindowsAzure.MobileServices.Sync
PushAction+<ExecuteBatchAsync>d__7.MoveNext ()

**Comment by MatkovIvan** [comment]
Hi @ronnyhash,

Thanks for getting in touch!
I've moved this line into the try/catch block (See #443), it should fix the described problem.
If this is time-critical for you, I recommend building the SDK from sources (with the fix) instead of waiting for a new nuget release.

Best,
Ivan

**Comment by ronnyhash** [comment]
Thank you for the quick response. Since this is a critical fix and since others are affected by it as well, when can we expect a new nuget release?

**Comment by ronnyhash** [comment]
Hi @MatkovIvan, what is the release schedule for this fix? It needs to come out in a nuget release.

**Comment by elamalani** [comment]
@ronnyhash We dont have a release planned right now. In the next few weeks, we will verify the SDK works with Android P and iOS 12 and we will include the fix with that release.

**Comment by ronnyhash** [comment]
Thank you @elamalani for the update! Having the release in the next few weeks will be great!

**Comment by chriscrown** [comment]
I got into this bug too last week. I'm pretty new in Azure Mobile Apps so I don't know if this helps:
While analyzing my client data I realized that the top most operation in the __operation table is always referencing a record (with the itemId) which doesn't exist on the client. Even when the operation kind is "update", the record to be updated is not present in the referenced table on the client. (I think this can also be caused by a bug in my code). The consequence of this missing reference is the broken sync described above.

Hope this helps!
